app/models/my_module.rb

```diff
@@ -500,6 +501,16 @@ def assign_user(user, assigned_by = nil)
 
   private
 
+  def set_completed
+    if my_module_status.final_status?
+      self.state = 'completed'
+    else
+      self.state = 'uncompleted'
+    end
+
+    set_completed_on
```

Comment:
Shouldn't it be called by 'before_validation :set_completed_on, if: :state_changed?'